Zend/zend_API.c

```diff
@@ -1846,6 +1846,39 @@ ZEND_API zend_result object_init_ex(zval *arg, zend_class_entry *class_type) /*
 }
 /* }}} */
 
+ZEND_API zend_result object_init_with_constructor(zval *arg, zend_class_entry *class_type, uint32_t param_count, zval *params, HashTable *named_params) /* {{{ */
+{
+	zend_result status = _object_and_properties_init(arg, class_type, NULL);
+	if (UNEXPECTED(status == FAILURE)) {
+		return FAILURE;
+	}
+	zend_object *obj = Z_OBJ_P(arg);
+	zend_function *constructor = obj->handlers->get_constructor(obj);
+	if (UNEXPECTED(constructor == NULL)) {
+		return FAILURE;
```

Comment:
Depending on the use-case of this function, I think that returning SUCCESS here when no arguments were passed, would be a better API. Otherwise we can not use this function as an internal equivalent of `new` without checking the existence of a constructor before that.

`ReflectionClass::newInstance()` does that, and throws `Class %s does not have a constructor, so you cannot pass any constructor arguments` when arguments where passed.

Reply:
This is actually a good point, really wondering if the SPL API that was exposed was not completely broken in reality now...

Reply:
I am wondering, should we throw an exception, or just allow it to initialize like a `new` call with too many arguments? As this is not currently an error in PHP: https://3v4l.org/CHps3, and it would be inconsistent with passing too many parameters which _would_ initialize it.

Reply:
In this case I would prefer to follow the behavior of `new`, then

Reply:
Actually I think we need to handle internal and userland classes differently.

Need to add a test for this when I come back to work on it.